lib/spack/spack/util/log_parse.py

```diff
@@ -0,0 +1,134 @@
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License (as
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from __future__ import print_function
+
+import re
+from six import StringIO
+
+from llnl.util.tty.color import colorize
+
+
+class LogEvent(object):
+    """Class representing interesting events (e.g., errors) in a build log."""
+    def __init__(self, text, line_no,
+                 pre_context='', post_context='', repeat_count=0):
+        self.text = text
+        self.line_no = line_no
+        self.pre_context = pre_context
+        self.post_context = post_context
+        self.repeat_count = repeat_count
+
+    @property
+    def start(self):
+        """First line in the log with text for the event or its context."""
+        return self.line_no - len(self.pre_context)
+
+    @property
+    def end(self):
+        """Last line in the log with text for event or its context."""
+        return self.line_no + len(self.post_context) + 1
+
+    def __getitem__(self, line_no):
+        """Index event text and context by actual line number in file."""
+        if line_no == self.line_no:
+            return self.text
+        elif line_no < self.line_no:
+            return self.pre_context[line_no - self.line_no]
+        elif line_no > self.line_no:
+            return self.post_context[line_no - self.line_no - 1]
+
+    def __str__(self):
+        """Returns event lines and context."""
+        out = StringIO()
+        for i in range(self.start, self.end):
+            if i == self.line_no:
+                out.write('  >> %-6d%s' % (i, self[i]))
+            else:
+                out.write('     %-6d%s' % (i, self[i]))
+        return out.getvalue()
+
+
+def parse_log_events(logfile, context=6):
+    """Extract interesting events from a log file as a list of LogEvent.
+
+    Args:
+        logfile (str): name of the build log to parse
+        context (int): lines of context to extract around each log event
+
+    Currently looks for lines that contain the string 'error', ignoring case.
+
+    TODO: Extract warnings and other events from the build log.
+    """
+    with open(logfile, 'r') as f:
+        lines = [line for line in f]
+
+    log_events = []
+    for i, line in enumerate(lines):
+        if re.search('error', line, re.IGNORECASE):
```

Comment:
I would highly recommend using `error:` instead of `error`. The latter will pick up hundreds of lines of the compiler linking to error-related modules.